RelationalRecord/src/Relations/User.hs

```diff
@@ -0,0 +1,80 @@
+{-# LANGUAGE TemplateHaskell, MultiParamTypeClasses, FlexibleInstances #-}
+{-# LANGUAGE RecordWildCards #-}
+
+module  Relations.User where
+
+import  Types.User as User
+import  Types.DB
+import  Relations.DB
+
+import  Database.Relational.Query
+import  Database.HDBC.Query.TH      (makeRecordPersistableDefault)
+
+import  Data.Int                    (Int32)
+
+
+
+-- SELECTS
+
+allUsers :: Relation () Users
+allUsers = User.users
+
+getUser :: Relation PKey Users
+getUser = relation' . placeholder $ \usrId -> do
+    a       <- query users
+    wheres  $ a ! User.id' .=. usrId
+    return  a
+
+
+-- INSERTS
+
+data UserInsert = UserInsert
+    { iTenantId     :: PKey
+    , iUsername     :: String
+    , iPassword     :: String
+    , iFirstName    :: Maybe String
+    , iLastName     :: Maybe String
+    }
+$(makeRecordPersistableDefault ''UserInsert)
+
+piUser :: Pi Users UserInsert
+piUser = UserInsert
+    |$| User.tenantId'
+    |*| User.username'
+    |*| User.password'
+    |*| User.firstName'
+    |*| User.lastName'
+
+insertUser :: Insert UserInsert
+insertUser = derivedInsert piUser
+
+
+-- UPDATES
+
+data UserUpdate = UserUpdate
+    { uTenantId     :: Maybe PKey
+    , uUsername     :: Maybe String
+    , uPassword     :: Maybe String
+    , uFirstName    :: Maybe (Maybe String)
+    , uLastName     :: Maybe (Maybe String)
+    , uStatus       :: Maybe Int32
+    }
+
+userUpdate :: UserUpdate
+userUpdate = UserUpdate
+    Nothing Nothing Nothing Nothing Nothing Nothing
+
+updateUserVariadic :: UserUpdate -> TimestampedUpdate
+updateUserVariadic UserUpdate {..} = derivedUpdate $ \projection -> do
+    User.tenantId'  <-#? uTenantId
+    User.username'  <-#? uUsername
+    User.password'  <-#? uPassword
+    User.firstName' <-#? uFirstName
+    User.lastName'  <-#? uLastName
+    User.status'    <-#? uStatus
+
+    (phTStamp, _)   <- placeholder (\tStamp -> User.updatedAt' <-# tStamp)
```

Comment:
The helper type for this update is ```type TimestampedUpdate = Update (DBTime, PKey)
```. This means, the generated query will have two placeholders, one for the timestamp, one for the primary key to update. The other updated values are rendered in the query as literals. So when rendering the query DSL, we get something like:
```UPDATE users SET first_name='Foo', last_name='Bar', updatedAt=? WHERE id=?```, where substitution of these placeholders happens one step later.

In HRR, these type parameters for e.g. an ```Update a``` have to be realized via the placeholder syntax. It's also possible to apply some datatype argument first, and use the literal inside the query, e.g. ```DBTime -> PKey -> Update ()```. One can decide on the context which is preferable.

Either way, getting a current timestamp from the system clock is done in the DBInterface, which in turn then runs the update query.